src/modules/roc_sndio/target_sox/roc_sndio/sox_backend.cpp

```diff
@@ -202,126 +238,72 @@ void SoxBackend::set_frame_size(core::nanoseconds_t frame_length,
-    core::Mutex::Lock lock(mutex_);
-
-    driver = map_to_sox_driver(driver);
-
-    if (!select_defaults(driver, inout)) {
-        return false;
-    }
-
-    const sox_format_handler_t* handler = sox_write_handler(inout, driver, NULL);
-    if (!handler) {
-        return false;
-    }
-
-    if (handler->flags & SOX_FILE_DEVICE) {
-        if ((filter_flags & FilterDevice) == 0) {
-            return false;
-        }
-    } else {
-        if ((filter_flags & FilterFile) == 0) {
-            return false;
-        }
-    }
-
-    return true;
-}
-
 ISink* SoxBackend::open_sink(core::IAllocator& allocator,
                              const char* driver,
                              const char* output,
-                             const Config& config) {
+                             const Config& config,
+                             int filter_flags) {
     core::Mutex::Lock lock(mutex_);
-
     first_created_ = true;
 
     driver = map_to_sox_driver(driver);
-
-    if (!select_defaults(driver, output)) {
-        return NULL;
-    }
-
     core::ScopedPtr<SoxSink> sink(new (allocator) SoxSink(allocator, config), allocator);
-    if (!sink) {
-        return NULL;
-    }
-
-    if (!sink->valid()) {
-        return NULL;
+    if (check_and_open(driver, output, filter_flags, sink)) {
+        return sink.release();
     }
-
-    if (!sink->open(driver, output)) {
-        return NULL;
-    }
-
-    return sink.release();
+    return NULL;
 }
 
 ISource* SoxBackend::open_source(core::IAllocator& allocator,
                                  const char* driver,
                                  const char* input,
-                                 const Config& config) {
+                                 const Config& config,
+                                 int filter_flags) {
     core::Mutex::Lock lock(mutex_);
 
     first_created_ = true;
 
     driver = map_to_sox_driver(driver);
-
-    if (!select_defaults(driver, input)) {
-        return NULL;
-    }
-
     core::ScopedPtr<SoxSource> source(new (allocator) SoxSource(allocator, config),
                                       allocator);
-    if (!source) {
-        return NULL;
+    if (check_and_open(driver, input, filter_flags, source)) {
+        return source.release();
     }
 
-    if (!source->valid()) {
-        return NULL;
-    }
-
-    if (!source->open(driver, input)) {
-        return NULL;
-    }
-
-    return source.release();
+    return NULL;
 }
 
-bool SoxBackend::get_drivers(core::StringList& list, int filter_flags) {
+bool SoxBackend::get_drivers(core::Array<DriverInfo>& list, int filter_flags) {
     core::Mutex::Lock lock(mutex_);
 
+    for (size_t n = 0; n < ROC_ARRAY_SIZE(default_drivers); n++) {
+        const char* driver = map_from_sox_driver(default_drivers[n]);
+        add_driver_mapping(list, driver, this,
+                           DriverDevice | DriverDefault | DriverSource | DriverSink);
+    }
```

Comment:
If format is not enabled is SoX (at compile time or as a plugin), we shouldn't add corresponding driver; but in this loop we're adding all drivers from default_drivers unconditionally.

We do we need a separate loop for default drivers? Can we use a single loop below and just add corresponding flag if is_default() reports true, instead of skipping driver?

Reply:
The reason I used a separate loop for default drivers is for the order of the priority in which they are supposed to be opened. If the loop below is used the format names from the handler are not returned in the same order. Without this order we will fallback to different drivers arbitrarily. Do you suggest an alternate method for this?

Reply:
Ah, I see! Let's keep it as is. Could you please add corresponding comment to the first loop?